File: project files/server/src/controllers/accountController.js
const asyncHandler = require('../utils/asyncHandler');
const accountService = require('../services/accountService');

const getProfile = asyncHandler(async (req, res) => {
  const data = await accountService.getProfile(req.auth.userId);
  res.status(200).json({ success: true, data });
});

const updateProfile = asyncHandler(async (req, res) => {
  const data = await accountService.updateProfile({ userId: req.auth.userId, ...req.body });
  res.status(200).json({ success: true, data });
});

const updateNotifications = asyncHandler(async (req, res) => {
  const data = await accountService.setNotificationPreferences({
    userId: req.auth.userId,
    notificationPreferences: req.body
  });
  res.status(200).json({ success: true, data });
});

const updateAddresses = asyncHandler(async (req, res) => {
  const data = await accountService.updateAddresses({ userId: req.auth.userId, addresses: req.body.addresses || [] });
  res.status(200).json({ success: true, data });
});

const listSessions = asyncHandler(async (req, res) => {
  const data = await accountService.listActiveSessions(req.auth.userId);
  res.status(200).json({ success: true, data });
});

const setTwoFactor = asyncHandler(async (req, res) => {
  const data = await accountService.setTwoFactor({
    userId: req.auth.userId,
    enabled: !!req.body.enabled
  });
  res.status(200).json({ success: true, data });
});

const listSavedCards = asyncHandler(async (req, res) => {
  const data = await accountService.listSavedCards(req.auth.userId);
  res.status(200).json({ success: true, data });
});

const addSavedCard = asyncHandler(async (req, res) => {
  const data = await accountService.addSavedCard({
    userId: req.auth.userId,
    card: req.body
  });
  res.status(200).json({ success: true, data });
});

const removeSavedCard = asyncHandler(async (req, res) => {
  const data = await accountService.removeSavedCard({
    userId: req.auth.userId,
    cardId: req.params.cardId
  });
  res.status(200).json({ success: true, data });
});

const setDefaultCard = asyncHandler(async (req, res) => {
  const data = await accountService.setDefaultCard({
    userId: req.auth.userId,
    cardId: req.params.cardId
  });
  res.status(200).json({ success: true, data });
});

const requestDataDownload = asyncHandler(async (req, res) => {
  await accountService.requestDataDownload(req.auth.userId);
  res.status(200).json({ success: true, message: 'Data download request submitted' });
});

const requestProfileOtp = asyncHandler(async (req, res) => {
  await accountService.requestProfileChangeOtp({
    userId: req.auth.userId,
    type: req.body.type,
    value: req.body.value
  });
  res.status(200).json({ success: true, message: 'OTP sent successfully' });
});

const verifyProfileOtp = asyncHandler(async (req, res) => {
  await accountService.verifyProfileChangeOtp({
    userId: req.auth.userId,
    type: req.body.type,
    value: req.body.value,
    otp: req.body.otp
  });
  res.status(200).json({ success: true, message: 'OTP verified successfully' });
});

const changePassword = asyncHandler(async (req, res) => {
  await accountService.changePassword({
    userId: req.auth.userId,
    currentPassword: req.body.currentPassword,
    newPassword: req.body.newPassword
  });
  res.status(200).json({ success: true, message: 'Password changed successfully' });
});

const logoutAllDevices = asyncHandler(async (req, res) => {
  await accountService.logoutAllDevices(req.auth.userId);
  res.status(200).json({ success: true, message: 'Logged out from all devices' });
});

const requestDeletion = asyncHandler(async (req, res) => {
  await accountService.requestAccountDeletion(req.auth.userId);
  res.status(200).json({ success: true, message: 'Account deletion requested' });
});

const addRecentlyViewed = asyncHandler(async (req, res) => {
  await accountService.addRecentlyViewed({
    userId: req.auth.userId,
    productId: req.body.productId
  });
  res.status(200).json({ success: true, message: 'Recently viewed updated' });
});

const listRecentlyViewed = asyncHandler(async (req, res) => {
  const data = await accountService.getRecentlyViewed(req.auth.userId);
  res.status(200).json({ success: true, data });
});

module.exports = {
  getProfile,
  updateProfile,
  updateNotifications,
  updateAddresses,
  listSessions,
  setTwoFactor,
  listSavedCards,
  addSavedCard,
  removeSavedCard,
  setDefaultCard,
  requestDataDownload,
  requestProfileOtp,
  verifyProfileOtp,
  changePassword,
  logoutAllDevices,
  requestDeletion,
  addRecentlyViewed,
  listRecentlyViewed
};

File: spring_back/src/main/java/com/groceryapp/controller/AccountController.java
package com.groceryapp.controller;

import com.groceryapp.security.AuthContext;
import com.groceryapp.service.AccountService;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/account")
public class AccountController {
    private final AccountService accountService;

    public AccountController(AccountService accountService) {
        this.accountService = accountService;
    }

    @GetMapping("/me")
    public Map<String, Object> getProfile() {
        return Map.of("success", true, "data", accountService.getProfile(AuthContext.current().getUserId()));
    }

    @PatchMapping("/me")
    public Map<String, Object> updateProfile(@RequestBody Map<String, Object> body) {
        return Map.of("success", true, "data", accountService.updateProfile(AuthContext.current().getUserId(), body));
    }

    @PatchMapping("/me/notifications")
    public Map<String, Object> updateNotifications(@RequestBody Map<String, Object> body) {
        return Map.of("success", true, "data", accountService.updateNotifications(AuthContext.current().getUserId(), body));
    }

    @PutMapping("/me/addresses")
    public Map<String, Object> updateAddresses(@RequestBody Map<String, Object> body) {
        return Map.of("success", true, "data", accountService.updateAddresses(AuthContext.current().getUserId(), castAddressRows(body.get("addresses"))));
    }

    @GetMapping("/me/sessions")
    public Map<String, Object> listSessions() {
        return Map.of("success", true, "data", accountService.listSessions(AuthContext.current().getUserId()));
    }

    @PatchMapping("/me/two-factor")
    public Map<String, Object> setTwoFactor(@RequestBody Map<String, Object> body) {
        return Map.of("success", true, "data", accountService.setTwoFactor(AuthContext.current().getUserId(), bool(body.get("enabled"))));
    }

    @GetMapping("/me/cards")
    public Map<String, Object> listSavedCards() {
        return Map.of("success", true, "data", accountService.listSavedCards(AuthContext.current().getUserId()));
    }

    @PostMapping("/me/cards")
    public Map<String, Object> addSavedCard(@RequestBody Map<String, Object> body) {
        return Map.of("success", true, "data", accountService.addSavedCard(AuthContext.current().getUserId(), body));
    }

    @DeleteMapping("/me/cards/{cardId}")
    public Map<String, Object> removeSavedCard(@PathVariable String cardId) {
        return Map.of("success", true, "data", accountService.removeSavedCard(AuthContext.current().getUserId(), cardId));
    }

    @PatchMapping("/me/cards/{cardId}/default")
    public Map<String, Object> setDefaultCard(@PathVariable String cardId) {
        return Map.of("success", true, "data", accountService.setDefaultCard(AuthContext.current().getUserId(), cardId));
    }

    @PostMapping("/me/change-password")
    public Map<String, Object> changePassword(@RequestBody Map<String, Object> body) {
        accountService.changePassword(
                AuthContext.current().getUserId(),
                str(body.get("currentPassword")),
                str(body.get("newPassword"))
        );
        return Map.of("success", true, "message", "Password changed successfully");
    }

    @PostMapping("/me/logout-all")
    public Map<String, Object> logoutAllDevices() {
        accountService.logoutAllDevices(AuthContext.current().getUserId());
        return Map.of("success", true, "message", "Logged out from all devices");
    }

    @PostMapping("/me/data-download")
    public Map<String, Object> requestDataDownload() {
        accountService.requestDataDownload(AuthContext.current().getUserId());
        return Map.of("success", true, "message", "Data download request submitted");
    }

    @PostMapping("/me/profile-otp/request")
    public Map<String, Object> requestProfileOtp(@RequestBody Map<String, Object> body) {
        accountService.requestProfileOtp(AuthContext.current().getUserId(), str(body.get("type")), str(body.get("value")));
        return Map.of("success", true, "message", "OTP sent successfully");
    }

    @PostMapping("/me/profile-otp/verify")
    public Map<String, Object> verifyProfileOtp(@RequestBody Map<String, Object> body) {
        accountService.verifyProfileOtp(
                AuthContext.current().getUserId(),
                str(body.get("type")),
                str(body.get("value")),
                str(body.get("otp"))
        );
        return Map.of("success", true, "message", "OTP verified successfully");
    }

    @PostMapping("/me/delete-request")
    public Map<String, Object> requestDeletion() {
        accountService.requestDeletion(AuthContext.current().getUserId());
        return Map.of("success", true, "message", "Account deletion requested");
    }

    @PostMapping("/me/recently-viewed")
    public Map<String, Object> addRecentlyViewed(@RequestBody Map<String, Object> body) {
        accountService.addRecentlyViewed(AuthContext.current().getUserId(), str(body.get("productId")));
        return Map.of("success", true, "message", "Recently viewed updated");
    }

    @GetMapping("/me/recently-viewed")
    public Map<String, Object> listRecentlyViewed() {
        return Map.of("success", true, "data", accountService.listRecentlyViewed(AuthContext.current().getUserId()));
    }

    @SuppressWarnings("unchecked")
    private List<Map<String, Object>> castAddressRows(Object v) {
        if (v instanceof List<?> list) {
            return (List<Map<String, Object>>) list;
        }
        return List.of();
    }

    private String str(Object v) {
        return v == null ? "" : String.valueOf(v).trim();
    }

    private boolean bool(Object v) {
        if (v instanceof Boolean b) return b;
        return "true".equalsIgnoreCase(str(v)) || "1".equals(str(v));
    }
}
